src/databricks/labs/ucx/source_code/python/python_analyzer.py

```diff
@@ -0,0 +1,280 @@
+            # nothing to do, dbutils.notebook.run uses a dedicated context
+            return InheritedContext(None, False)
+        if isinstance(base_node, MagicLine):
+            return base_node.build_inherited_context(self._context, child_path)
+        logger.warning(f"Can't build inherited context for node {NodeBase.__name__} of type {type(base_node).__name__}")
+        return InheritedContext(None, False)
+
+    def _register_import(self, base_node: ImportSource) -> Iterable[DependencyProblem]:
+        prefix = ""
+        if isinstance(base_node.node, ImportFrom) and base_node.node.level is not None:
+            prefix = "." * base_node.node.level
+        name = base_node.name or ""
+        problems = self._context.parent.register_import(prefix + name)
+        for problem in problems:
+            prob = self._filter_import_problem_in_try_except(problem, base_node)
+            if prob is not None:
+                yield prob
+
+    @classmethod
+    def _filter_import_problem_in_try_except(
+        cls, problem: DependencyProblem, base_node: ImportSource
+    ) -> DependencyProblem | None:
+        if problem.code != 'import-not-found':
+            return problem
+        # is base_node in a try-except clause ?
+        node = base_node.node.parent
+        while node and not isinstance(node, Try):
+            node = node.parent
+        if cls._is_try_except_import_error(node):
+            return None
+        return problem
+
+    @classmethod
+    def _is_try_except_import_error(cls, node: Try | None) -> bool:
+        if not isinstance(node, Try):
+            return False
+        for handler in node.handlers:
+            if isinstance(handler.type, Name):
+                if handler.type.name == "ImportError":
+                    return True
+        return False
+
+    def _register_notebook(self, base_node: NotebookRunCall) -> Iterable[DependencyProblem]:
+        has_unresolved, paths = base_node.get_notebook_paths(self._context.session_state)
+        if has_unresolved:
+            yield DependencyProblem(
+                'dependency-cannot-compute-value',
+                f"Can't check dependency from {base_node.node.as_string()} because the expression cannot be computed",
+            )
+        for path in paths:
+            # notebooks ran via dbutils.notebook.run do not inherit or propagate context
+            yield from self._context.parent.register_notebook(Path(path), False)
+
+    def _mutate_path_lookup(self, change: SysPathChange) -> Iterable[DependencyProblem]:
+        if isinstance(change, UnresolvedPath):
+            yield DependencyProblem(
+                'sys-path-cannot-compute-value',
+                f"Can't update sys.path from {change.node.as_string()} because the expression cannot be computed",
+            )
+            return
+        change.apply_to(self._context.path_lookup)
+
+
+T = TypeVar("T")
+
+
+class MagicLine(NodeBase):
```

Comment:
magic line abstraction belongs to notebook and works for all languages, not just python

Reply:
High level I agree, but this one is Python only (it inherits NodeBase), and it's there to avoid an cyclic dependency (cells.py imports python_analyzer.py, which then can't import cells.py...). Or should I move MagicLine under a dedicated file in notebooks package ?

Reply:
magics don't belong to python (only).

Reply:
moved to notebooks/magic.py